easybuild/tools/containers.py

```diff
@@ -0,0 +1,304 @@
+# Copyright 2017-2018 Ghent University
+#
+# This file is part of EasyBuild,
+# originally created by the HPC team of Ghent University (http://ugent.be/hpc/en),
+# with support of Ghent University (http://ugent.be/hpc),
+# the Flemish Supercomputer Centre (VSC) (https://www.vscentrum.be),
+# Flemish Research Foundation (FWO) (http://www.fwo.be/en)
+# and the Department of Economy, Science and Innovation (EWI) (http://www.ewi-vlaanderen.be/en).
+#
+# https://github.com/easybuilders/easybuild
+#
+# EasyBuild is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation v2.
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Support for generating container recipes and creating container images
+
+:author: Shahzeb Siddiqui (Pfizer)
+:author: Kenneth Hoste (HPC-UGent)
+"""
+import os
+from distutils.version import LooseVersion
+from vsc.utils import fancylogger
+
+from easybuild.tools.build_log import EasyBuildError, print_msg
+from easybuild.tools.config import CONT_IMAGE_FORMAT_EXT3, CONT_IMAGE_FORMAT_SANDBOX, CONT_IMAGE_FORMAT_SQUASHFS
+from easybuild.tools.config import build_option, container_path
+from easybuild.tools.filetools import which, write_file
+from easybuild.tools.module_naming_scheme.utilities import det_full_ec_version
+from easybuild.tools.run import run_cmd
+
+
+CONT_TYPE_DOCKER = 'docker'
+CONT_TYPE_SINGULARITY = 'singularity'
+
+DOCKER = 'docker'
+LOCALIMAGE = 'localimage'
+SHUB = 'shub'
+SINGULARITY_BOOTSTRAP_TYPES = [DOCKER, LOCALIMAGE, SHUB]
+
+
+SINGULARITY_TEMPLATE = """
+Bootstrap: %(bootstrap)s
+From: %(from)s
+
+%%post
+%(install_os_deps)s
+
+# upgrade easybuild package automatically to latest version
+pip install -U easybuild
```

Comment:
I guess this is up for debate, and part of the reason why we're putting this under `--experimental` for now.

The worry was that if we don't update EasyBuild, you'll be stuck to a particular EasyBuild version in this container, which can be a problem if you'll be using the container image you create as a base for others.

I guess we could consider making this optional, and having some other mechanism (like allowing to specify a particular EasyBuild version to use in the container).

For now, I'd leave this as is, we can revisit this later?

Reply:
@boegel  i agree with you having the `pip install -U easybuild` in the container recipe is better than having to update or support many easybuild base container. It would be overkill to create a base container for every easybuild version but rather let pip install the latest version inside the container so hopefully the eb on the host and inside the host are of the same version.